### DISCUSSION TITLE
Release 18 milestone

### DISCUSSION THREAD

**Discussion by AlmasB** [gluonhq/scenebuilder#525] [General]
If we're going by the java/javafx release cycle, then the next release should be end of March / beginning of April. This means we have 4-5 weeks to go. If everyone is OK with this, I'd suggest we wrap up outstanding [issues](https://github.com/gluonhq/scenebuilder/issues?q=is%3Aopen+is%3Aissue+milestone%3A18) and PRs if we can, and re-target to 19 if we can't. All new PRs (unless pure `test:` or `docs:`) should now probably target 19.

As soon as javafx 18 is out, we should update and get some users to help thoroughly test SB 18.

**Comment by johanvos** [comment]
I agree outstanding PR's and issues should now target 19.
I recommend following the OpenJFX approach: create a branch for stabilisation of SB 18 while work on 19 can continue in the main branch.
Testing can be done with JavaFX 18-ea+11 which should be really close to the final 18 which will be released next month.

**Comment by Oliver-Loeffler** [comment]
I'll see that I can finish #427 next days. Not certain, If I can finish #396 this week but I'll invest some time, unfortunately this one got bigger than expected.

**Comment by johanvos** [comment]
@AlmasB @Oliver-Loeffler @aalmiray Are you ok with branching sb18 now? If so, I will do that, and someone probably has to change the actions so that 18-builds are created from the sb18 branch.

From that moment on, the 18-branch should only be used to fix showstoppers and critical issues. All new functionality or bug fixes that are not critical (preventing SB to work) should go in main.

**Reply by AlmasB** [reply]
I think we should:

1. update main to `JavaFX 18-ea+11`
2. Re-tag early-access for this as 18-ea, so that the release is not overwritten by rolling releases, for testing purposes.
3. branch away to `SB18` and leave it until javafx 18 release, then update the branch to javafx 18 and release.
4. continue developing on main as normal

**Reply by abhinayagarwal** [reply]
This sounds like a good plan. Did we ever get to a point of branching out `SB18` ? Now that JavaFX 18 is available we should branch off from a commit which we are most comfortable with.

**Reply by AlmasB** [reply]
yes, `sb18` branch needs updating to the stable javafx 18 and we can release.

**Reply by abhinayagarwal** [reply]
In addition to JavaFX 18, we need to wait until Java 18 is released.

**Comment by aalmiray** [comment]
@johanvos @AlmasB the `release.yml` workflow may be triggered form any branch. The `early-access.yml` workflow only reacts to changes pushed to `master` branch.

**Comment by AlmasB** [comment]
> I think we should:
> 
> 1. update main to `JavaFX 18-ea+12`
> 2. Re-tag early-access for this as 18-ea, so that the release is not overwritten by rolling releases, for testing purposes.
> 3. branch away to `SB18` and leave it until javafx 18 release, then update the branch to javafx 18 and release.
> 4. continue developing on main as normal

All of the above are now done. `sb18` branch is there for stabilisation and can be used for release in due time. The PRs can now continue to be accepted to `master`.

**Reply by aalmiray** [reply]
You may also configure a different `<jreleaser><project><snapshot><pattern>` instead of changing the project version. The default pattern is set to `*-SNAPSHOT`.

**Comment by AlmasB** [comment]
@abhinayagarwal I believe sb18 is now ready for release. We should probably keep it java 17 for a while, given we just switched to it. Based on the previous releases, perhaps we could keep java 17 until the next LTS release.

**Reply by abhinayagarwal** [reply]
We always ship SB with latest Java and JavaFX releases.

**Reply by AlmasB** [reply]
ok, so we need a PR to update java version to 18?

**Reply by jperedadnr** [reply]
yes, we need to update all workflows to use Java/JavaFX 18. I guess for both sb18 and master, right?

**Reply by jperedadnr** [reply]
PR https://github.com/gluonhq/scenebuilder/pull/534 for sb18

**Reply by jperedadnr** [reply]
Hmm, no temurin 18?
We should use `openjdk`, shouldn't we?

**Reply by jperedadnr** [reply]
trying jdk.java.net...

**Reply by jperedadnr** [reply]
https://github.com/gluonhq/scenebuilder/pull/534 is green now

**Reply by AlmasB** [reply]
> I guess for both sb18 and master, right?

yes, I suppose we can just re-PR back to master

**Reply by jperedadnr** [reply]
ok, merged.
Not sure if we need anything else to do the release? Just a PR removing the `-SNAPSHOT`?

**Reply by AlmasB** [reply]
Yes, I think just removing snapshot should do it, `sb18` does not have anything extra from the moment we decided to freeze it.

We haven't run jreleaser on a non-snapshot version, so I don't know if the changelog will be fine, but I can fix it manually post-release.

**Reply by jperedadnr** [reply]
Done: https://github.com/gluonhq/scenebuilder/pull/536

**Reply by jperedadnr** [reply]
merged, let's see...

**Reply by jperedadnr** [reply]
hmm, nothing is running?

**Reply by aalmiray** [reply]
1. update version in pom.xml to remove snapshot
2. update the release.yml to use JDK 18
3. run the Release workflow from the UI

**Reply by aalmiray** [reply]
> hmm, nothing is running?

Because the release workflow requires a manual trigger.

**Reply by jperedadnr** [reply]
1 ✅ 
2 ✅ 
3 I'm hitting it now:
<img width="356" alt="image" src="https://user-images.githubusercontent.com/2043230/160396044-f8cc331f-0738-413e-9361-26bdd97645ba.png">

**Reply by aalmiray** [reply]
In the future, updating the pom.xml for release may also happen automatically when running the workflow. If you'd like this to happen let me know and I'll send a PR

**Reply by jperedadnr** [reply]
seems that is failing now:
https://github.com/gluonhq/scenebuilder/blob/sb18/.github/workflows/release.yml#L40

```
Failed to execute goal org.codehaus.mojo:versions-maven-plugin:2.10.0:set (default-cli) on project parent: You must specify the new version, either by using the newVersion property (that is -DnewVersion=... on the command line) or run in interactive mode
```

**Reply by AlmasB** [reply]
> mvn -B versions:set -DremoveSnapshot

We manually set version 18.0.0 in the poms, should we not have done this?

**Reply by jperedadnr** [reply]
Maybe we need to rollback the last commit that removed the snapshot? 
```
mvn -B versions:set -DremoveSnapshot
```
already takes care of it, and fails if there is no snapshot...

**Reply by aalmiray** [reply]
Ah. So it's already taken care of. Yes, rollback is needed.

**Reply by jperedadnr** [reply]
Ok, let's try again...

**Reply by jperedadnr** [reply]
weird, it is not picking the changes I just did for the java setup 🤔

**Reply by jperedadnr** [reply]
The repository is correctly checked out from sb18. However, it looks like the workflow is still from master:
```
uses: gluonhq/scenebuilder/.github/workflows/bundles-windows.yml@master
```
...

**Reply by jperedadnr** [reply]
Hopefully this should do it: https://github.com/gluonhq/scenebuilder/pull/540

**Reply by aalmiray** [reply]
Hmm looks like reusable workflow refs would need updates as well. Btw. push the updates on master, then pass the java version as param like before

**Reply by jperedadnr** [reply]
hmm, thad doesn't work:
<img width="593" alt="image" src="https://user-images.githubusercontent.com/2043230/160403731-fe8aea78-9d58-4208-9b9d-d81687cf5e4b.png">

**Reply by jperedadnr** [reply]
will this work?
```
uses: gluonhq/scenebuilder/.github/workflows/bundles-linux.yml@sb18
```
Or should I rollback it to `@master`, and then apply the changes (related to setup java actions) to `master` as Andrés suggests?

**Reply by AlmasB** [reply]
Perhaps we can apply all changes to master and release from master?

**Reply by jperedadnr** [reply]
Hmm, what do we have sb18 for then? There are no relevant changes anyway

**Reply by aalmiray** [reply]
I'd suggest making master the release branch, use feature branches for dev

**Reply by jperedadnr** [reply]
So we only need this commit from sb18 in master?
https://github.com/gluonhq/scenebuilder/commit/a38204b14cbbdad7813a378c01fce36a7c4fc47c

**Reply by aalmiray** [reply]
No. You also need the changes in the `release.yml` workflow

**Reply by jperedadnr** [reply]
the changes are already included

**Reply by AlmasB** [reply]
For the future, alternatively, we can have two branches to work for the general case:

* master (maybe rename to dev) - for development
* release - for release.

The release.yml workflow targets only the release branch (`@release` if this works). That way we don't need to constantly update branch names as we go through versions 19, 20 etc. We will still need to update java/fx versions but that's a different issue. Once we need to freeze the release, we dump `dev` to `release` and wait for the java/fx versions to be released. Once they do, we need to somehow PR both branches with the new version, release on `release` and continue working on `dev`.

**Reply by jperedadnr** [reply]
Ok, PR against master: https://github.com/gluonhq/scenebuilder/pull/542

**Reply by jperedadnr** [reply]
In any case, this PR: https://github.com/gluonhq/scenebuilder/pull/544 rolls back the change done to release workflow, so once #542 is merged, we can still release from  'sb18`.

**Reply by AlmasB** [reply]
The release seems to be stuck on:

```
          mvn -B -ntp \
            -Drepository.url=https://${{ github.actor }}:${{ secrets.GITHUB_TOKEN }}@github.com/${{ github.repository }}.git \
            release:prepare
          mvn -pl :parent process-resources
```

Should `secrets.GITHUB_TOKEN` be `secrets.PAT`? Just a guess...

**Reply by jperedadnr** [reply]
no idea....
@aalmiray can you guess what's going on?

**Reply by aalmiray** [reply]
I think that `release:prepare` can be avoided if the version has been set already. SB does not publish JARs to Maven Central.

**Reply by jperedadnr** [reply]
Ok, then we cancel the workflow. Could  you create a PR with Athena required charge? Thx.

**Comment by aalmiray** [comment]
AFAICT `${{ secrets.GITHUB_TOKEN }}` is enough when making commits to the same repository where the GH workflow is run, which is the case for the release workflow.

**Comment by aalmiray** [comment]
FWIW @jperedadnr your commit message does not follow the Conventional Commits spec

**Reply by jperedadnr** [reply]
Which one? Can you elaborate?

**Reply by aalmiray** [reply]
```
commit 981b2650fca45911b04837b435ca40fcac78a9de (HEAD -> release-update, origin/master, origin/HEAD, master)
Author: José Pereda <pereda@eii.uva.es>
Date:   Tue Mar 29 18:25:54 2022 +0200

    Use Java/JavaFX version 18 (#542)
    
    Co-authored-by: jose.pereda <jose.pereda@gluonhq.com>
```

See https://github.com/gluonhq/scenebuilder/blob/master/CONTRIBUTING.md#pull-request-standards

**Reply by jperedadnr** [reply]
I added `build` to the PR title only, not to the commit message, I see.

**Comment by aalmiray** [comment]
PR is ready for review https://github.com/gluonhq/scenebuilder/pull/545

**Reply by jperedadnr** [reply]
Once we merge this, we should run `release` workflow from master, right? There is no point in having  `sb18` branch anymore then...

**Comment by jperedadnr** [comment]
@aalmiray @AlmasB I believe `master` branch is ready for starting the `release` workflow from the master branch (and not from `sb18`). Any comment? Else I'll start it.

**Reply by aalmiray** [reply]
I'd say, go for it!

**Comment by jperedadnr** [comment]
hmm
https://github.com/gluonhq/scenebuilder/runs/5754075429?check_suite_focus=true#step:6:75

```

remote: error: GH006: Protected branch update failed for refs/heads/master.        
remote: error: At least 1 approving review is required by reviewers with write access. 
```

**Reply by AlmasB** [reply]
```
mvn -B -ntp versions:set versions:commit -DnewVersion=${{ needs.precheck.outputs.PROJECT_VERSION }}
          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
          git config --global user.name "GitHub Action"
          git commit -a -m "Releasing version ${{ needs.precheck.outputs.PROJECT_VERSION }}"
          git push origin master
```

this won't work because the master branch is protected

**Reply by AlmasB** [reply]
regardless of how the version is set, there shouldn't be any pushes to master, or it will fail. I wonder if we can manually update the version and skip mvn versions altogether?

**Reply by jperedadnr** [reply]
We did that before for sb18, removing `-SNAPSHOT` with a PR.
Should we do the same on master? That would be a single commit, then we'll do the release workflow, then another manual commit to put the snapshot back...

**Reply by AlmasB** [reply]
This sounds like it could work. I think the main goal now is for us to complete one successful release cycle (i.e. release 18), so we know what's involved and what the new process looks like. Then we can reflect and refine the process, as needed.

So, yeah, let's try removing `mvn versions` from the workflow and manually set the version to `18.0.0`.

**Reply by jperedadnr** [reply]
okay, since @aalmiray did the previous PR for the versions change, probably he can do this change (remove for now the Versions step), and then we can do the final PR to remove the SNAPSHOT?

**Reply by aalmiray** [reply]
Will take me some hrs before I'm back in front of a keyboard.

**Reply by aalmiray** [reply]
You can remove the commit step but keep the version change (from latest PR) .. OR .. expect master to have non snapshot version but that would probably require updating the reusable workflows (though they themselves also do not push any commits).

**Reply by jperedadnr** [reply]
We have in the precheck job:

- Set environment
```
mvn -B versions:set -DremoveSnapshot
```
That will fail if we remove the SNAPSHOT manually (as we already tested in sb18).

But it actually does all the builds with the release version, even if nothing is committed to master.

Then we have the release job, it checks out again the repository, so I take it gets again the SNAPSHOT.

I wonder if the `Version` step should just do the same and instead of committing the new version it should simply call again:

`mvn -B versions:set -DremoveSnapshot`

Would that work for JReleaser?

For a later step, we should set the new SNAPSHOT version (18.0.1-SNAPSHOT, 18.1.0-SNAPSHOT or 19.0.0-SNAPSHOT...), once the release finishes successfully.

**Reply by jperedadnr** [reply]
Let's see: https://github.com/gluonhq/scenebuilder/pull/547

**Reply by jperedadnr** [reply]
Feel free to approve so we can test (at this point another failed workflow doesn't make a difference ;) )

**Reply by AlmasB** [reply]
Does GH allow the PR author approving their own PRs? If so, perhaps, we should do that for now -- for the purposes of an expedient solution.

**Reply by jperedadnr** [reply]
Usually you can approve yourself, unless there are restrictions added to the repository (like the need of having at least one approval)

**Reply by jperedadnr** [reply]
Merged, let's see how it goes now...

**Comment by jperedadnr** [comment]
hmm 
https://github.com/gluonhq/scenebuilder/runs/5760399516?check_suite_focus=true#step:7:31
@aalmiray , help!

```
Error: s3.app.region must not be blank. Configure a value using the Maven DSL, or define a JRELEASER_S3_APP_REGION environment variable, or define a key/value pair in /home/runner/.jreleaser/config.properties with key = JRELEASER_S3_APP_REGION
Error:  s3.app.bucket must not be blank. Configure a value using the Maven DSL, or define a JRELEASER_S3_APP_BUCKET environment variable, or define a key/value pair in /home/runner/.jreleaser/config.properties with key = JRELEASER_S3_APP_BUCKET
```

Workflow has `JRELEASER_S3_GLUON_*`, not `JRELEASER_S3_APP_*`??

**Reply by jperedadnr** [reply]
@aalmiray @AlmasB any idea? should we just try to rename `JRELEASER_S3_GLUON_*` to `JRELEASER_S3_APP_*`?

**Reply by aalmiray** [reply]
Edit `pom.xml` lines 242 & 247, change from app to gluon.

**Reply by jperedadnr** [reply]
like this?
https://github.com/gluonhq/scenebuilder/pull/548

**Reply by aalmiray** [reply]
Correct

**Reply by jperedadnr** [reply]
Great, maybe @AlmasB can approve then

**Reply by jperedadnr** [reply]
@aalmiray It failed again: https://github.com/gluonhq/scenebuilder/runs/5775002841?check_suite_focus=true#step:7:24
```
release failed: Path does not exist. target/settings.properties
```
Any idea?

**Reply by aalmiray** [reply]
Gah... now I remember why the `mvn process-resources` goal was used before, to generate this file plus another one. It must be added back. Look at the recent changes and make sure it's invoked after setting the version.

**Reply by AlmasB** [reply]
So, something like this?

```
      - name: Set Release Version
        run: |
          mvn -B -ntp -q versions:set -DnewVersion=${{ needs.precheck.outputs.PROJECT_VERSION }} -DgenerateBackupPoms=false
          mvn -pl :parent process-resources
```

**Reply by jperedadnr** [reply]
I had that same change, indeed: https://github.com/gluonhq/scenebuilder/pull/549

**Reply by jperedadnr** [reply]
@aalmiray one more:
https://github.com/gluonhq/scenebuilder/runs/5775661417?check_suite_focus=true#step:7:56

```
release failed: The specified bucket is not valid. (Service: Amazon S3; Status Code: 400; Error Code: InvalidBucketName; Request ID: TKA2ZSEX5JFRG9AA; S3 Extended Request ID: 8RbOEB/sG9gC+XsTHQdgs2Y6rnoHQPWa6AqksMjNJv35HkuFpVBUquxw3ATCWWLVo4zWaFphHee29O/WqBnoYA==; Proxy: null)
```

**Reply by aalmiray** [reply]
I'll have a look later today. In the meantime inspect the jreleaser logs (trace.log) added as workflow artifact

**Reply by AlmasB** [reply]
The last line in the trace.log file is: `checking if bucket s3://download.gluonhq.com exists`. That plus the error message above seem to indicate something to do with s3...

**Reply by aalmiray** [reply]
According to the log the `s3` section has the following configuration

```
upload:
    enabled: true
    s3:
        gluon:
            enabled: true
            active: release
            connectTimeout: 20
            readTimeout: 60
            artifacts: true
            files: true
            signatures: false
            checksums: false
            region: us-east-1
            bucket: s3://download.gluonhq.com
            accessKeyId: ************
            secretKey: ************
            sessionToken: **unset**
            path: scenebuilder/{{artifactFile}}
```

The configured bucket name (as coming from the secrets) includes the `s3://` protocol and a custom domain. This won't work as is.
I'm assuming there's an `*.amazonaws.com` URL that can be used to upload SB artifacts, something similar to `https://{{bucket}}.s3.{{region}}.amazonaws.com`. If this is the case then we just have to update the value for `bucket`.

**Reply by aalmiray** [reply]
OTOH if a custom S3 endpoint (such as `download.gluonhq.com`) should be used then please configure a value for the `<endpoint>` property but be sure to also configure a value for `<downloadUrl>` and `<bucket>` properties. See https://jreleaser.org/guide/latest/configuration/upload/s3.html for reference (click on the Maven tab).

**Reply by jperedadnr** [reply]
We have renamed the bucket to `download.gluonhq.com`, trying again now.

**Reply by jperedadnr** [reply]
Looks like we have a winner! 
GitHub releases and S3 have the artifacts!

**Reply by jperedadnr** [reply]
FYI: I've also built Scene Builder on my M1, and pushed the binary to the release assets. In case anyone can test it. 

(I already have a self hosted runner to build on M1, but needs some wiring to the release process..., we can do that later)

**Comment by AlmasB** [comment]
Right, just to "close" this discussion, 18 is now out (many thanks @aalmiray @jperedadnr ). I've fixed the legacy changelog commits, so they work with the new format, but the next version shouldn't have these issues.
